### PR TITLE
Overlap Plotting

### PR DESCRIPTION
**Pull request by pshriwise**
This add the option to plot overlaps in geometry. Users can specify they want geometry overlaps to appear using a `<show_overlaps>true</show_overlaps>` element and specify a color for the overlap using `<overlap_color>`. This is particularly useful in voxel plots to identify and isolate geometry errors.

I'm wondering if this might be an appropriate default behavior for plotting and would be happy to update the code to make it so if that's what we decide is best.

### CODE REVIEW

File: include/openmc/plot.h [hunk 1]
```diff
@@ -106,6 +112,7 @@ class PlotBase {
   Position width_; //!< Plot width in geometry
   PlotBasis basis_; //!< Plot basis (XY/XZ/YZ)
   std::array<size_t, 3> pixels_; //!< Plot size in pixels
+  bool color_overlaps_;
```
Comment (paulromano) [inline include/openmc/plot.h]:
```suggestion
  bool color_overlaps_; //!< Show overlapping cells?
```

File: include/openmc/geometry.h [hunk 2]
```diff
@@ -47,7 +47,7 @@ inline bool coincident(double d1, double d2) {
 //! Check for overlapping cells at a particle's position.
 //==============================================================================
 
-bool check_cell_overlap(Particle* p);
+ bool check_cell_overlap(Particle* p, bool error=true);
```
Comment (paulromano) [inline include/openmc/geometry.h]:
```suggestion
bool check_cell_overlap(Particle* p, bool error=true);
```

File: tests/regression_tests/plot_overlaps/test.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    """Specialized TestHarness for running OpenMC plotting tests."""
+    def __init__(self, plot_names):
+        super().__init__(None)
+        self._plot_names = plot_names
+
+    def _run_openmc(self):
+        openmc.plot_geometry(openmc_exec=config['exe'])
+
+
+
```
Comment (paulromano) [inline tests/regression_tests/plot_overlaps/test.py]: PEP8: One blank line between methods

File: src/plot.cpp [hunk 4]
```diff
@@ -619,6 +638,38 @@ Plot::set_mask(pugi::xml_node plot_node)
   }
 }
 
+void Plot::set_overlap_color(pugi::xml_node plot_node) {
+  color_overlaps_ = false;
+  if (check_for_node(plot_node, "show_overlaps")) {
+    color_overlaps_ = get_node_value_bool(plot_node, "show_overlaps");
+    overlap_color_ = RED;
```
Comment (paulromano) [inline src/plot.cpp]: Probably should set this as default in the class definition itself?

Comment (pshriwise) [reply src/plot.cpp]: Definitely. Good call.

File: openmc/plots.py [hunk 5]
```diff
@@ -391,6 +405,25 @@ def mask_background(self, mask_background):
...
+    def overlap_color(self, overlap_color):
+        cv.check_type('plot overlap color', overlap_color, Iterable)
+        if isinstance(overlap_color, str):
+            if overlap_color.lower() not in _SVG_COLORS:
+                raise ValueError("'{}' is not a valid color.".format(overlap_color))
+        else:
+            cv.check_length('plot overlap color', overlap_color, 3)
+            for rgb in overlap_color:
+                cv.check_greater_than('plot overlap color', rgb, 0, True)
+                cv.check_less_than('plot overlap color', rgb, 256)
```
[the remark this answers is not in the document]
Comment (pshriwise) [reply openmc/plots.py]: Definitely!

File: openmc/plots.py [hunk 6]
```diff
@@ -430,6 +429,19 @@ def meshlines(self, meshlines):
 
         self._meshlines = meshlines
 
+    @staticmethod
+    def check_color(err_string, color):
```
Comment (paulromano) [inline openmc/plots.py]: Make this `_check_color` since a user is not expected to be calling this directly

**Comment by paulromano** [comment]
Looks like there's one more place you can use `check_color` in the setter for `meshlines`

**Comment by pshriwise** [comment]
Nearly forgot to add the overlap property to the capi `PlotBase`!

**Comment by pshriwise** [comment]
Nearly forgot to add the overlap property to the `PlotBase` class.

**Pull request closed without merging**